kernel/thread.c

```diff
@@ -273,12 +273,14 @@ static size_t copy_bytes(char *dest, size_t dest_size, const char *src, size_t s
 const char *k_thread_state_str(k_tid_t thread_id, char *buf, size_t buf_size)
 {
 	size_t      off = 0;
-	uint8_t     bit;
 	uint8_t     thread_state = thread_id->base.thread_state;
-	static const char  *states_str[8] = {"dummy", "pending", "prestart",
-					     "dead", "suspended", "aborting",
-					     "", "queued"};
-	static const size_t states_sz[8] = {5, 7, 8, 4, 9, 8, 0, 6};
+	static const uint8_t state[6] = {_THREAD_DUMMY, _THREAD_PENDING,
+					 _THREAD_PRESTART, _THREAD_DEAD,
+					 _THREAD_SUSPENDED, _THREAD_QUEUED};
+	static const char  *states_str[6] = {"dummy", "pending",
+					     "prestart", "dead",
+					     "suspended", "queued"};
+	static const size_t states_sz[6] = {5, 7, 8, 4, 9, 6};
```

Comment:
Since you're in here reworking the logic anyway:  this is storing 9 bytes for each record (for the state value, pointer and the size), in addition to the string bytes themselves, which are never less than five bytes ("dead\0").  So just give it a 10 byte array (long enough for "suspended\0") indexed by the bit value (which is tight packed), and you actually save memory.  And the expression is simpler too as it doesn't require manually sorting or (yikes) calculating strlen() results.  Something like:  (note it does need an expression for the bit index of the states, right now all we have is masks in the headers)

````c
static const char states[][12] = {
    [DUMMY_BIT] = "dummy",
    [PENDING_BIT] = "pending",
    [PRESTART_BIT] = "prestart",
    [DEAD_BIT] = "dead",
    [SUSPENDED_BIT] = "suspended",
    [ABORTING_BIT] = "aborting",
    [QUEUED_BIT] = "queued",
};
````

The rest of the logic doesn't need to change beyond the obvious pointer/array swap.  Also the copy_bytes() can just turn into a strncpy or whatever since you don't need the size value any more.